sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala

```diff
@@ -259,6 +259,156 @@ class StatisticsSuite extends QueryTest with TestHiveSingleton with SQLTestUtils
+      StatsSetupConst.NUM_PARTITIONS,
+      StatsSetupConst.ROW_COUNT,
+      StatsSetupConst.RAW_DATA_SIZE,
+      StatsSetupConst.TOTAL_SIZE).forall(!tableMetadata.contains(_)))
+
+    assert(StringUtils.filterPattern(describeResult1, "*numRows\\s+500*").nonEmpty)
+    checkStats(
+      tabName, isDataSourceTable = false, hasSizeInBytes = true, expectedRowCounts = Some(500))
+  }
+
+  test("alter table rename after analyze table") {
+    val oldName = "tab1"
+    val newName = "tab2"
+    withTable(oldName, newName) {
+      val fetchedStats1 = createNonPartitionedTable(oldName)
+      sql(s"ALTER TABLE $oldName RENAME TO $newName")
+      val fetchedStats2 = checkStats(
+        newName, isDataSourceTable = false, hasSizeInBytes = true, expectedRowCounts = Some(500))
+      assert(fetchedStats1 == fetchedStats2)
+
+      // ALTER TABLE RENAME does not affect the contents of Hive specific statistics
+      val hiveClient = spark.sharedState.externalCatalog.asInstanceOf[HiveExternalCatalog].client
+      val describeResult = hiveClient.runSqlHive(s"DESCRIBE FORMATTED $newName")
+      assert(StringUtils.filterPattern(describeResult, "*numRows\\s+500*").nonEmpty)
+    }
+  }
+
+  test("alter table SET TBLPROPERTIES after analyze table") {
+    val tabName = "tab1"
+    withTable(tabName) {
+      val fetchedStats1 = createNonPartitionedTable(tabName)
+
+      sql(s"ALTER TABLE $tabName SET TBLPROPERTIES ('foo' = 'a')")
+      val fetchedStats2 = checkStats(
+        tabName, isDataSourceTable = false, hasSizeInBytes = true, expectedRowCounts = Some(500))
+      assert(fetchedStats1 == fetchedStats2)
+
+      // ALTER TABLE SET TBLPROPERTIES invalidates some contents of Hive specific statistics
+      // This is triggered by the Hive alterTable API
+      val hiveClient = spark.sharedState.externalCatalog.asInstanceOf[HiveExternalCatalog].client
+      val describeResult = hiveClient.runSqlHive(s"DESCRIBE FORMATTED $tabName")
+      assert(StringUtils.filterPattern(describeResult, "*numRows\\s+-1*").nonEmpty)
+    }
+  }
+
+  test("alter table UNSET TBLPROPERTIES after analyze table") {
+    val tabName = "tab1"
+    withTable(tabName) {
+      val fetchedStats1 = createNonPartitionedTable(tabName)
+
+      sql(s"ALTER TABLE $tabName UNSET TBLPROPERTIES ('prop1')")
+      val fetchedStats2 = checkStats(
+        tabName, isDataSourceTable = false, hasSizeInBytes = true, expectedRowCounts = Some(500))
+      assert(fetchedStats1 == fetchedStats2)
+
+      // ALTER TABLE UNSET TBLPROPERTIES invalidates some contents of Hive specific statistics
+      // This is triggered by the Hive alterTable API
+      val hiveClient = spark.sharedState.externalCatalog.asInstanceOf[HiveExternalCatalog].client
+      val describeResult = hiveClient.runSqlHive(s"DESCRIBE FORMATTED $tabName")
+      assert(StringUtils.filterPattern(describeResult, "*numRows\\s+-1*").nonEmpty)
+    }
+  }
+
+  test("add/drop partitions - managed table") {
```

Comment:
FYI, when we drop partitions of EXTERNAL tables, `ANALYZE TABLE` is unable to exclude them from statistics. This should be fixed with https://issues.apache.org/jira/browse/SPARK-17129, if my understanding is right.